### ISSUE TITLE
Incorrect comparison result for "equal" values of different float types

### ISSUE THREAD

**Issue by philrz** [brimdata/super#5295]
## tl;dr

As a user, this result seems incorrect to me.

```
$ echo '0.99 (float32)' | zq -Z 'yield this > 0.99' -
true
```

## Details

Repro is with Zed commit 2357e17.

This came up while trying to reproduce the equivalent of the [mgbench bench1 q4](https://github.com/andrewcrotty/mgbench/blob/master/bench1/q4/q4.sql) in Zed.

Compare the three queries below.

```
$ zq -version
Version: v1.17.0-74-g2357e178

$ echo '0.99' | zq -Z 'yield this > 0.99' -
false

$ echo '0.99 (float32)' | zq -Z 'yield this > 0.99' -
true

$ echo '0.99 (float32)' | zq -Z 'yield this > float32(0.99)' -
false
```

I know floats are notoriously fussy when it comes to establishing strict equality between values, so I assume the explanation has something to do with that.

This came up because the [mgbench bench1 DDL](https://github.com/andrewcrotty/mgbench/blob/master/bench1/ddl/ddl.sql) puts the values into 32-bit floats and then the [query](https://github.com/andrewcrotty/mgbench/blob/master/bench1/q4/q4.sql) applies  a filter `cpu_wio > 0.99`. To attempt the equivalent in Zed, I shaped the values to `float32` at load time and applied a similar filter in the Zed query, but the problem described here meant several `cpu_wio` equal to `0.99` made it past the filter and were counted in the query, effectively producing an incorrect result for the benchmark. I can certainly work around this for now by adding the `float32(0.99)` cast, but I'm not sure if this is something we'd expect users to do.

**Comment by philrz** [comment]
I've come to find that ClickHouse shows what looks like the same problem. Like SuperDB, the problem seems to stem from the fact that "bare" values like `0.99` are 64-bit floats that don't compare reliably with 32-bit float equivalents.

```
$ clickhouse-client
ClickHouse client version 24.12.2.29 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 24.12.2.

Warnings:
 * Maximum number of threads is lower than 30000. There could be problems with handling a lot of simultaneous queries.

wpm.local :) SELECT toTypeName(0.99);

SELECT toTypeName(0.99)

Query id: 4a2f1318-10e0-40c0-afb8-01fe9bc2d37a

   ┌─toTypeName(0.99)─┐
1. │ Float64          │
   └──────────────────┘

1 row in set. Elapsed: 0.001 sec. 

wpm.local :) SELECT 0.99 > 0.99;

SELECT 0.99 > 0.99

Query id: 0f6a4063-2e8e-4ed6-bae1-a3fca67d1386

   ┌─greater(0.99, 0.99)─┐
1. │                   0 │
   └─────────────────────┘

1 row in set. Elapsed: 0.001 sec. 

wpm.local :) SELECT CAST(0.99 AS Float32) > 0.99;

SELECT CAST(0.99, 'Float32') > 0.99

Query id: 8a0fdd91-7a81-4090-903e-c18ce8002197

   ┌─greater(CAST(0.99, 'Float32'), 0.99)─┐
1. │                                    1 │
   └──────────────────────────────────────┘

1 row in set. Elapsed: 0.001 sec. 

wpm.local :) SELECT CAST(0.99 AS Float32) > CAST(0.99 AS Float32);

SELECT CAST(0.99, 'Float32') > CAST(0.99, 'Float32')

Query id: 878b079d-3c22-4826-9ef5-846948aff637

   ┌─greater(CAST(0.99, 'Float32'), CAST(0.99, 'Float32'))─┐
1. │                                                     0 │
   └───────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec. 
```

Meanwhile, DuckDB does _not_ show the problem with similar repro steps, seemingly because their "bare" values like `0.99` are `DECIMAL` type.

```
$ duckdb
v1.1.3 19864453f7
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D SELECT TYPEOF(0.99);
┌──────────────┐
│ typeof(0.99) │
│   varchar    │
├──────────────┤
│ DECIMAL(3,2) │
└──────────────┘
D SELECT 0.99 > 0.99;
┌───────────────┐
│ (0.99 > 0.99) │
│    boolean    │
├───────────────┤
│ false         │
└───────────────┘
D SELECT CAST(0.99 AS Float) > 0.99;
┌──────────────────────────────┐
│ (CAST(0.99 AS FLOAT) > 0.99) │
│           boolean            │
├──────────────────────────────┤
│ false                        │
└──────────────────────────────┘
D SELECT CAST(0.99 AS Float) > CAST(0.99 AS Float);
┌─────────────────────────────────────────────┐
│ (CAST(0.99 AS FLOAT) > CAST(0.99 AS FLOAT)) │
│                   boolean                   │
├─────────────────────────────────────────────┤
│ false                                       │
└─────────────────────────────────────────────┘
```

**Comment by philrz** [comment]
We discussed this one as a team and reached consensus that any approaches with compared float types that would hide this effect from users are undesirable. Especially given that other established systems like ClickHouse also live with this effect, we're considering it an acceptable hazard when users choose to work with floating point. Offering a "decimal" type that doesn't have the effect and maybe using it by default for fractional values like we see with DuckDB is something we'll definitely consider in future development as user demand dictates, so that's being tracked in #1522. Closing this issue.
